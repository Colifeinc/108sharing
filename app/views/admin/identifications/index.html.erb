<% set_meta_tags title: I18n.t("activerecord.models.identification") %>
<% set_meta_tags og: { title: I18n.t("activerecord.models.identification") } %>

<%= react_component("pages/admin/identifications/index",
  { identifications: @identifications_json },
  prerender: false) %>
