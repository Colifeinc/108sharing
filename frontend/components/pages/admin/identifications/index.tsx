import * as React from 'react'
import I18n from '../../../../core/i18n'
import { IWindow } from '../../../../core/interfaces'
import { IJsonResponse } from '../../../../core/JsonApiSerializer'
import { adminService } from '../../../../core/services'
import { Button } from '../../../atoms'
import AdminLayout from 'components/layouts/AdminLayout'
import { EditTable } from 'components/organisms'

declare let window: IWindow

interface IProps {
  identifications: {
    identifications: IJsonResponse
  }
}

const COLUMNS = [
  {
    name: I18n.t('generic.user'),
    field: record => record.user.username,
  },
  {
    name: I18n.t('generic.image'),
    field: record => (
      <div className="flex items-center flex-wrap my-2 p-4 bg-neutral-100">
        {record.identification_images.map(image => (
          <div key={image.id} className="w-1/2 p-4 max-w-64">
            <a
              href={image.image_url}
              target="_blank"
              className="block transition duration-150 hover:opacity-85"
            >
              <img key={image.id} src={image.image_url} alt="" />
            </a>
          </div>
        ))}
      </div>
    ),
  },
  {
    name: I18n.t('generic.status'),
    field: record => I18n.t(`enums.identification.workflow_state.${record.workflow_state}`),
  },
]

const AdminIdentificationsIndex: React.FC<IProps> = props => {
  const { identifications: initialIdentifications } = adminService.getIdentificationsFromJson(
    props.identifications.identifications
  )
  const [identifications, setIdentifications] = React.useState(initialIdentifications)
  const filteredIdentifications = React.useMemo(() => {
    return identifications.filter(identification => {
      return identification.workflow_state === 'requested'
    })
  }, [identifications])

  const approveIdentification = React.useCallback(
    async identification => {
      const {
        identification: approvedIdentification,
        flush,
      } = await adminService.approveIdentification(identification.id)
      setIdentifications(
        identifications.map(item =>
          item.id === approvedIdentification.id ? approvedIdentification : { ...item }
        )
      )
      if (window && window.flashMessages) {
        window.flashMessages.addMessage({ text: flush.message, type: flush.type })
      }
    },
    [identifications]
  )

  const declineIdentification = React.useCallback(
    async identification => {
      const {
        identification: declinedIdentification,
        flush,
      } = await adminService.declineIdentification(identification.id)
      setIdentifications(
        identifications.map(item =>
          item.id === declinedIdentification.id ? declinedIdentification : { ...item }
        )
      )
      if (window && window.flashMessages) {
        window.flashMessages.addMessage({ text: flush.message, type: flush.type })
      }
    },
    [identifications]
  )

  const actionColumn = {
    name: I18n.t('generic.action'),
    field: record => (
      <>
        <Button type="button" size="small" handleClick={() => approveIdentification(record)}>
          {I18n.t('generic.approve')}
        </Button>
        <Button
          className="mt-3"
          type="button"
          buttonType="muted"
          size="small"
          handleClick={() => declineIdentification(record)}
        >
          {I18n.t('generic.decline')}
        </Button>
      </>
    ),
  }

  return (
    <AdminLayout model="identification" title={I18n.t('activerecord.models.identification')}>
      <EditTable
        editable={false}
        columns={[...COLUMNS, actionColumn]}
        records={filteredIdentifications}
      />
    </AdminLayout>
  )
}

export default AdminIdentificationsIndex
