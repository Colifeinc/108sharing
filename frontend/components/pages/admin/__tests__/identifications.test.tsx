// tslint:disable-next-line
import '../../../../test/matchMedia.mock'

import { act } from '@testing-library/react'
import { mount, shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import * as React from 'react'
import AdminIdentificationIndex from 'components/pages/admin/identifications/index'
import { identificationsJsonWithRequestedUser } from '../../../../test/fixtures/identifications'

jest.mock('../../../../core/services/AdminService')
jest.mock('../../../../core/i18n')

let container

beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  document.body.removeChild(container)
  container = null
})

describe('AdminIdentificationIndex', () => {
  const props = {
    identifications: {
      identifications: identificationsJsonWithRequestedUser,
    },
  }

  it('should matches to the snapshot', () => {
    const wrapper = shallow(<AdminIdentificationIndex {...props} />)

    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should remove identification request when approveIdentification has clicked', async () => {
    let wrapper

    act(() => {
      wrapper = mount(<AdminIdentificationIndex {...props} />, { attachTo: container })
    })

    expect(wrapper.find('tbody').find('tr')).toHaveLength(2)

    await act(async () => {
      wrapper
        .find('button[type="button"]')
        .first()
        .simulate('click')
    })
    wrapper.update()

    expect(wrapper.find('tbody').find('tr')).toHaveLength(1)
  })

  it('should remove identification request when declineIdentification has clicked', async () => {
    let wrapper

    act(() => {
      wrapper = mount(<AdminIdentificationIndex {...props} />, { attachTo: container })
    })

    expect(wrapper.find('tbody').find('tr')).toHaveLength(2)

    await act(async () => {
      wrapper
        .find('button[type="button"]')
        .last()
        .simulate('click')
    })
    wrapper.update()

    expect(wrapper.find('tbody').find('tr')).toHaveLength(1)
  })
})
