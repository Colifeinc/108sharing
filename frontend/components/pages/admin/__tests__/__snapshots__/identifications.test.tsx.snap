// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminIdentificationIndex should matches to the snapshot 1`] = `
<AdminLayout
  model="identification"
  title="Identification"
>
  <EditTable
    columns={
      Array [
        Object {
          "field": [Function],
          "name": "User",
        },
        Object {
          "field": [Function],
          "name": "Image",
        },
        Object {
          "field": [Function],
          "name": "Status",
        },
        Object {
          "field": [Function],
          "name": "Action",
        },
      ]
    }
    editable={false}
    records={
      Array [
        Object {
          "description": null,
          "id": 1,
          "identification_images": Array [
            Object {
              "id": 1,
              "image_url": "/url/1",
            },
          ],
          "user": Object {
            "bio": "はじめまして。太郎と申します。都内に住んでおりますので、何かございましたらいつでもご連絡ください。すぐに対応いたします。",
            "email": "user@example.com",
            "email_notification": true,
            "id": 2,
            "posts_count": 3,
            "unread_notifications_count": 2,
            "username": "Taro",
          },
          "workflow_state": "requested",
        },
        Object {
          "description": null,
          "id": 2,
          "identification_images": Array [
            Object {
              "id": 2,
              "image_url": "/url/2",
            },
            Object {
              "id": 3,
              "image_url": "/url/3",
            },
          ],
          "user": Object {
            "bio": "はじめまして。花子です。",
            "email": "user@example.com",
            "email_notification": true,
            "id": 3,
            "posts_count": 4,
            "unread_notifications_count": 2,
            "username": "Hanako",
          },
          "workflow_state": "requested",
        },
      ]
    }
  />
</AdminLayout>
`;
